engine/resources/pippet_run.go

```diff
@@ -0,0 +1,160 @@
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package resources
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"os/exec"
+	"sync"
+)
+
+// PippetRunner is the interface to the wrapped "puppet yamlresource receive"
+// process that keeps running alongside mgmt. Individual resources are expected
+// to register themselves, and can use the Apply function from that point on.
+type PippetRunner interface {
+	Register(interface{}) error
+	Unregister(interface{}) error
+	Apply(string, string, string) PippetResult
+	Close() error
+}
+
+// PippetResult is the structured return value type for the PippetRunner's
+// Apply function.
+type PippetResult struct {
+	Error     bool
+	Failed    bool
+	Changed   bool
+	Exception string
+}
+
+type pippetReceiver struct {
+	stdin         io.WriteCloser
+	stdout        io.ReadCloser
+	registerMutex sync.Mutex
+	applyMutex    sync.Mutex
+	registered    int
+}
+
+var pippetRunnerInstance PippetRunner
+var pippetRunnerOnce sync.Once
+
+// PippetRunnerInstance returns a pointer to the PippetRunner implementation.
+// The PippetRunner is supposed to be a singleton object. The pippet resource
+// code should always use the PippetRunnerInstance function to gain access to the
+// pippetReceiver object. Other objects of type pippetReceiver should not be created.
+func PippetRunnerInstance() (PippetRunner, error) {
+	if pippetRunnerInstance == nil {
+		pippetRunnerOnce.Do(func() { pippetRunnerInstance = &pippetReceiver{} })
+	}
+	return pippetRunnerInstance, nil
+}
+
+// Init runs the Puppet process that will perform the work of synchronizing resources
+// that are sent to its stdin. The process will keep running until Close is called.
+// Init should not be called directly. It is implicitly called by the Register function.
+func (obj *pippetReceiver) Init() error {
+	cmd := exec.Command("puppet", "yamlresource", "receive", "--color=no")
+	var err error
+	obj.stdin, err = cmd.StdinPipe()
+	if err != nil {
+		return err
+	}
+	obj.stdout, err = cmd.StdoutPipe()
+	if err != nil {
+		obj.stdin.Close()
+		return err
+	}
+	if err := cmd.Start(); err != nil {
+		obj.stdin.Close()
+		return err
+	}
+	buf := make([]byte, 80)
+	if _, err := obj.stdout.Read(buf); err != nil {
+		obj.stdin.Close()
+		return err
+	}
+	return nil
+}
+
+// Register should be called by any user (i.e., any pippet resource) before using
+// the Apply function on this receiver object.
+// Register implicitly takes care of calling Init if required.
+func (obj *pippetReceiver) Register(res interface{}) error {
```

Comment:
The input isn't even used in this function... ?

Reply:
True, it's just flavor now - the resource registers "itself" even though the runner object uses a simple counter.
I removed it.